### ISSUE TITLE
Route KNX telegrams to USB interface and IP network

### ISSUE THREAD

**Issue by tosate** [knxd/knxd#392]
Hi there,

I am using knxd version 0.14.25 as a software interface that connects homebridge and openHAB to my KNX bus over a hager USB interface. This works pretty well.

Since a couple of days I have some Gosund SP1 smart sockets which I have flashed with the [Sonoff-Tasmota firmware.](https://github.com/arendst/Sonoff-Tasmota) The devices are connected to the Wi-Fi and support the KNX protocol now.

Unfortunately, I cannot toggle them over KNX. If I use the hardware switch or the web interface, I can see in the group monitor that they send KNX telegrams indicating their status.

My knxd configuration looks like this:
`/usr/lib/knxd_args -e 0.0.255 -E 0.0.2:50 -D -T -R -S -b usb:`

I am sending KNX telegrams with the knxtool like this:
`knxtool off ip:127.0.0.1 1/1/112`

Maybe those telegrams are routed to the KNX USB interface only and not to the IP network?
How can I debug this to find out whether this is the case?
Is there any magical configuration parameter that enables the routing of KNX telegrams to the USB interface and the IP network?
Do you have any other idea how I could fix this issue?

Cheers,
Thomas

**Comment by dallmair** [comment]
AFAICT, HomeBridge uses knxd's proprietary protocol, so if that one works we know that knxd itself is running and routing telegrams.

Again AFAICT, Sonoff-Tasmota uses IP-Routing (multicast). So the question is: Does IP-Routing not work at all in knxd, or is there some network device in your environment which is incapable of handling multicast communication? My guess is it's the latter, esp. because Wifi is involved. Fortunately, you have another component around (openHAB), which helps a lot.

How did you configure openHAB, via IP-Tunneling or IP-Routing? Does the openHAB system have a wired connection to knxd?
* If you're using IP-Routing and a wired connection to openHAB, most likely the Wifi access point/router needs to be reconfigured or replaced to make the new devices work.
* If you're using IP-Routing and a wireless connection to openHAB, I have no idea what could make the other devices fail.
* If you're using IP-Tunneling in openHAB, can you try to switch to IP-Routing, at least temporarily?

**Comment by tosate** [comment]
Thank you very much for your suggestions.
I have checked out the configurations of the services and collected some data.

HomeBridge uses knxd's TCP-Socket running on port 6720.

My openHAB instance runs on the same Raspberry Pi as knxd and it connects using the IP-Tunnel. Basically how it's described [here](https://www.openhab.org/addons/bindings/knx/).
I can try to switch to IP-Routing to check whether it works with knxd and there will be no other network device involved since the communication stays on the same machine.

The Sonoff-Tasmota devices are connected over UniFi AC Lite access points from Ubiquiti Networks. These devices should support the multicast communication. In fact, I can see a lot of multicast traffic there, since this is how the UniFi controller manages the access points. As I already wrote, I even receive the KNX status telegrams from the Sonoff-Tasmota devices. The only thing which seems not to work is sending KNX telegrams to them.

**Comment by tosate** [comment]
Up to now I wasn't able to connect openHAB over IP-Routing. There seems to be something wrong with multicast routing. I have to further investigate to come up with more specific details.

**Comment by tru7** [comment]
I don't think you can have more than one process communicating with a specific multicast address on a single OS.

**Comment by tosate** [comment]
Hi Othmar,

I think you have to elaborate this a bit more. As I have described before, HomeBridge connects over the TCP-Socket. openHAB uses the TUNNEL interface. Changing to ROUTING does not work. Which other process might already communicate over multicast?

**Comment by smurfix** [comment]
Yes you can. Both programs need to set a specific option that tells the multicast packets to loopback, and there are addressing issues (which knxd can work around, but some recipients refuse to accept the packets correctly if you do that), but in principle it works.

**Comment by tru7** [comment]
@tosate your knxd does with -R
@smurfix if it really works it's beyond my knowledge

**Comment by smurfix** [comment]
@tru7 That's the multi-port option (--multi-port / --single-port) on the multicast router, plus the `IP_MULTICAST_LOOP` socket option (`IPPROTO_IP`). See `src/libserver/eibnetip.cpp`.

Using this mechanism it's possible to run more than one knxd on the same host and let them talk to each other with multicast, which kindof implies that other programs can do that too if they do it right. The knxd test suite does that.

**Comment by tosate** [comment]
@tru7 my knxd configuration consists of the -R parameter

I have no idea why it's not working. The commmand netstat -g even displays the multicast group membership for the IPv4 address 224.0.23.12.

In the meantime I have managed to integrate the smart sockets in homebridge using MQTT.
However, it would be nice to integrate them over KNX because this would enable me to turn them on or off over the light switches.

**Comment by tru7** [comment]
@tosate yes I know you configured knxd to use multicast.  And I understand you want openHAB to use IP routing.
@smurfix confirmed that knxd is specifically written so two different knxd processes can share the same local multicast port. Did you check that openHAB is written the same special way? Because you want knxd and openHAB to share the same local multicast port.

**Comment by tosate** [comment]
I did some research yesterday evening and I have learned that I confused broadcast with multicast. In broadcast packets are sent to everyone whereas in multicast they are sent to subscribers of the multicast group.
I have implemented a small Java application to subscribe to the KNX multicast group as described in [here](
https://www.baeldung.com/java-broadcast-multicast).
After starting the application I could see in WireShark that my machine receives the packets from knxd whenever an event occurs.
I could even configure openHAB to use IP ROUTE. 

Unfortunately, I still cannot switch the smart sockets, but this will become an issue of the Sonoff-Tasmota project. You can close this issue.

**Comment by smurfix** [comment]
You can close it yourself. ;-)

**Comment by Noschvie** [comment]
> 
> 
> Unfortunately, I still cannot switch the smart sockets, but this will become an issue of the Sonoff-Tasmota project.

@tosate : did you get this solved with Tasmota ?
Thanks!

**Comment by tosate** [comment]
Sorry for the late answer. I had the impression that the Tasmota KNX module is not widely used. So I gave up on it. Getting rid of openHAB has also been he right decision.
Today I am running the following docker containers:
- homebridge
- mosquitto
- homeassistant

The smart sockets use MQTT and mosquitto is their broker. I have MQTT plug-ins in home assistant and homebridge to integrate them. However, when I have started to use the smart sockets, I wanted to be able to turn them on and off with my KNX switches. Therefore, I have created automations in homeassistant. They listen on the KNX bus for certain addresses and certain events. If someone uses the switch, the automation toggles the smart socket over MQTT and publishes the smart socket state on the KNX bus.
Maybe this setup is a bit complicated, but it works very reliable.
